src/hotspot/share/runtime/thread.cpp

```diff
@@ -2593,20 +2593,22 @@ int JavaThread::java_suspend_self() {
 // Helper routine to set up the correct thread state before calling java_suspend_self.
 // This is called when regular thread-state transition helpers can't be used because
 // we can be in various states, in particular _thread_in_native_trans.
-// Because this thread is external suspended the safepoint code will count it as at
-// a safepoint, regardless of what its actual current thread-state is. But
-// is_ext_suspend_completed() may be waiting to see a thread transition from
-// _thread_in_native_trans to _thread_blocked. So we set the thread state directly
-// to _thread_blocked. The problem with setting thread state directly is that a
+// The correct thread state of a suspended thread is _thread_blocked. With that state
+// safepoint/handshake code will count it as safepoint/handshake safe. Also it allows
+// another thread to continue if it is waiting in is_ext_suspend_completed() for this
+// thread to change state from _thread_in_native_trans to the target state.
```

Comment:
The revised wording doesn't really convey the situation to me. We _have_ to set the thread-state to _thread_blocked so that a thread waiting in _is_ext_suspend_completed can proceed (there is no generic "target" state - it must be _thread_blocked). I would simplify and rephrase as follows:

"We have to set the thread state directly to _thread_blocked so that it will be seen to be safepoint/handshake safe whilst suspended. This is also necessary to allow a thread in is_ext_suspend_completed, that observed the _thread_in_native_trans state, to proceed."

Reply:
Ok, I will use your version.